### PR TITLE
Consolidated Security Updates for 2.0.12 - 2.1.x

### PR DESCRIPTION
**Pull request by nickjjzhao**
Consolidated Security Updates for 2.0.12

- Apply three-strikes rule to all transaction failures
- Apply unconditional subjective CPU check along with some additional logging
- Provide options to enable subjective CPU billing for P2P and API transactions ,and provide an option to disable it for individual accounts

v2.0.x - #10264

Co-Authored-By:Kevin Heifner heifnerk@objectcomputing.com
Co-Authored-By:Lin Huang lin.huang@block.one

## Change Type
**Select *ONE*:**
- [ ] Documentation

- [ ] Stability bug fix

- [x] Other

- [ ] Other - special case

## Testing Changes
**Select *ANY* that apply:**
- [x] New Tests

- [ ] Existing Tests

- [ ] Test Framework

- [ ] CI System

### CODE REVIEW

**Review by heifner — DISMISSED** [review]

File: plugins/producer_plugin/producer_plugin.cpp [hunk 1]
```diff
@@ -444,15 +446,16 @@ class producer_plugin_impl : public std::enable_shared_from_this<producer_plugin
...
             if( future.valid() ) {
                future.wait();
-               app().post( priority::low, [self, future{std::move(future)}, persist_until_expired, next{std::move( next )}, trx_id]() mutable {
-                  auto exception_handler = [&next, trx_id](fc::exception_ptr ex) {
-                    fc_dlog(_trx_failed_trace_log, "[TRX_TRACE] Speculative execution is REJECTING tx: ${txid} : ${why} ",
-                            ("txid", trx_id)("why",ex->what()));
+               app().post( priority::low, [self, future{std::move(future)}, persist_until_expired, next{std::move( next )}, trx{std::move(trx)}]() mutable {
+                  auto exception_handler = [&next, trx{std::move(trx)}](fc::exception_ptr ex) {
+                     fc_dlog(_trx_successful_trace_log, "[TRX_TRACE] Speculative execution is REJECTING tx: ${txid}, auth: ${a} : ${why} ",
+                            ("txid", trx->id())("a",trx->get_transaction().first_authorizer())("why",ex->what()));
```
Comment (heifner) [inline plugins/producer_plugin/producer_plugin.cpp:456]: Remove the `_trx_successful_trace_log` logging here. In 2.1.x these were separated.

File: plugins/producer_plugin/producer_plugin.cpp [hunk 2]
```diff
@@ -479,26 +482,40 @@ class producer_plugin_impl : public std::enable_shared_from_this<producer_plugin
             if (std::holds_alternative<fc::exception_ptr>(response)) {
                _transaction_ack_channel.publish(priority::low, std::pair<fc::exception_ptr, transaction_metadata_ptr>(std::get<fc::exception_ptr>(response), trx));
                if (_pending_block_mode == pending_block_mode::producing) {
-                  fc_dlog(_trx_failed_trace_log, "[TRX_TRACE] Block ${block_num} for producer ${prod} is REJECTING tx: ${txid} : ${why} ",
+                  fc_dlog(_trx_successful_trace_log, "[TRX_TRACE] Block ${block_num} for producer ${prod} is REJECTING tx: ${txid}, auth: ${a} : ${why} ",
```
Comment (heifner) [inline plugins/producer_plugin/producer_plugin.cpp:485]: Remove the `_trx_successful_trace_log` logging here. In 2.1.x these were separated.

File: plugins/producer_plugin/producer_plugin.cpp [hunk 3]
```diff
@@ -479,26 +482,40 @@ class producer_plugin_impl : public std::enable_shared_from_this<producer_plugin
...
+                        ("why",std::get<fc::exception_ptr>(response)->what()));
+                  fc_dlog(_trx_failed_trace_log, "[TRX_TRACE] Block ${block_num} for producer ${prod} is REJECTING tx: ${txid}, auth: ${a} : ${why} ",
+                        ("block_num", chain.head_block_num() + 1)
+                        ("prod", get_pending_block_producer())
+                        ("txid", trx->id())
+                        ("a", trx->packed_trx()->get_transaction().first_authorizer())
                         ("why",std::get<fc::exception_ptr>(response)->what()));
                } else {
-                  fc_dlog(_trx_failed_trace_log, "[TRX_TRACE] Speculative execution is REJECTING tx: ${txid} : ${why} ",
+                  fc_dlog(_trx_successful_trace_log, "[TRX_TRACE] Speculative execution is REJECTING tx: ${txid}, auth: ${a} : ${why} ",
```
Comment (heifner) [inline plugins/producer_plugin/producer_plugin.cpp:498]: Remove the `_trx_successful_trace_log` logging here. In 2.1.x these were separated.

**Review by heifner — CHANGES_REQUESTED** [review]

File: plugins/producer_plugin/producer_plugin.cpp (hunk 1 above)
Comment (nickjjzhao) [reply plugins/producer_plugin/producer_plugin.cpp]: Fixed.

File: plugins/producer_plugin/producer_plugin.cpp (hunk 2 above)
Comment (nickjjzhao) [reply plugins/producer_plugin/producer_plugin.cpp]: Fixed.

File: plugins/producer_plugin/producer_plugin.cpp (hunk 3 above)
Comment (nickjjzhao) [reply plugins/producer_plugin/producer_plugin.cpp]: Fixed.

File: plugins/producer_plugin/producer_plugin.cpp [hunk 4]
```diff
@@ -1808,7 +1867,7 @@ bool producer_plugin_impl::process_unapplied_trxs( const fc::time_point& deadlin
             }
 
             auto prev_billed_cpu_time_us = trx->billed_cpu_time_us;
-            if( !_subjective_billing.is_disabled() && prev_billed_cpu_time_us > 0 && !rl.is_unlimited_cpu( first_auth )) {
```
Comment (heifner) [inline plugins/producer_plugin/producer_plugin.cpp:1811]: This should not have been removed. Also needs to be added to develop branch PR.

**Review by heifner — APPROVED** [review]

File: plugins/producer_plugin/producer_plugin.cpp (hunk 4 above)
Comment (nickjjzhao) [reply plugins/producer_plugin/producer_plugin.cpp:1811]: Fixed.

**Pull request merged**